ir/instr.cpp

```diff
@@ -2227,6 +2227,88 @@ unique_ptr<Instr> Memcpy::dup(const string &suffix) const {
 }
 
 
+vector<Value*> Strlen::operands() const {
+  return { ptr };
+}
+
+void Strlen::rauw(const Value &what, Value &with) {
+  RAUW(ptr);
+}
+
+void Strlen::print(ostream &os) const {
+  os << getName() << " = strlen " << *ptr;
+}
+
+
+// Given (bid, ofs0), encode an expression that returns the next offset ofs2
+// that has a null character.
+// Returns: <found?, ofs2>
+static pair<expr, expr>
+nextNull(const Memory &m, const expr &bid, const expr &ofs0, unsigned i,
+         const string &nextnull_uf) {
+  auto ofs = ofs0 + expr::mkUInt(i, ofs0.bits());
+
+  if (i == strlen_unroll_cnt) {
+    auto ofs2 = expr::mkUF(nextnull_uf, {bid, ofs}, ofs);
+    Pointer ptr2(m, bid, ofs2);
+    Byte b = m.load(ptr2);
+    // Approximate the situation by assuming that there is no null between
+    // ofs and ofs2
+    return { ptr2.inbounds(false, true) && !b.is_ptr() && !b.is_poison(false) &&
+             b.is_zero(), move(ofs2) };
+  }
+
+  Pointer ptr(m, bid, ofs);
+  Byte b = m.load(ptr);
+
+  auto is_char = !b.is_ptr() && !b.is_poison(false);
+  auto strictly_inbounds = ptr.inbounds(false, true);
+  if (is_char.isFalse() || strictly_inbounds.isFalse())
+    return { expr(false), ofs };
+  else if (is_char.isTrue() && b.is_zero().isTrue() &&
+           strictly_inbounds.isTrue())
+    return { expr(true), ofs };
+
+  auto [found, ofs_next] = nextNull(m, bid, ofs0, i + 1, nextnull_uf);
+  return { is_char && strictly_inbounds && (b.is_zero() || found),
+           expr::mkIf(b.is_zero(), ofs, ofs_next) };
+}
+
+StateValue Strlen::toSMT(State &s) const {
+  auto &[eptr, np_ptr] = s[*ptr];
+  s.addUB(np_ptr);
+
+  const auto &m = s.getMemory();
+  Pointer p(m, eptr);
+
+  string nextnull_uf = "nextnull_" + m.unique_id();
+  expr bid = expr::mkVar("bid", p.get_bid());
+  expr ofs = expr::mkVar("ofs", p.get_offset());
+  s.addAxiom(
+    expr::mkForAll({ bid, ofs },
+                   expr::mkUF(nextnull_uf, {bid, ofs}, ofs).uge(ofs)));
+
+  auto [found, ofs2] = nextNull(m, p.get_bid(), p.get_offset(), 0, nextnull_uf);
+
+  s.addUB(found);
```

Comment:
this should be addPre(), since we want to ignore executions where strlen isn't unrolled enough.

Reply:
What about a string that does not end with null character?
In the case, strlen should raise UB (https://en.cppreference.com/w/cpp/string/byte/strlen )